File: 백준/Gold/14943. 벼룩 시장/벼룩 시장.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int N = Integer.parseInt(br.readLine());
        int[] nums = new int[N];
        ArrayList<int[]> plus = new ArrayList<>();
        ArrayList<int[]> minus = new ArrayList<>();
        long result = 0;

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            nums[i] = Integer.parseInt(st.nextToken());
        }

        for (int i = 0; i < N; i++) {
            if (nums[i] > 0) {
                plus.add(new int[]{i, nums[i]});
            } else if (nums[i] < 0) {
                minus.add(new int[]{i, nums[i]});
            }
        }

        int m = 0;
        int p = 0;
        while (m < minus.size()) {
            int mIdx = minus.get(m)[0];
            int minusNum = minus.get(m)[1];
            m++;

            while (minusNum < 0) {
                int pIdx = plus.get(p)[0];
                int plusNum = plus.get(p)[1];

                if (minusNum + plusNum > 0) {
                    result += (long) -minusNum * Math.abs(pIdx - mIdx);
                    plus.set(p, new int[]{pIdx, plusNum+minusNum});
                    minusNum = 0;
                } else {
                    result += (long) plusNum * Math.abs(pIdx - mIdx);
                    minusNum += plusNum;
                    p++;
                }
            }
        }

        System.out.println(result);
    }
}

File: 백준/Gold/14943. 벼룩 시장/벼룩 시장.py
# 벼룩 시장
import sys
input = sys.stdin.readline

N = int(input())
nums = list(map(int,input().split()))
plus = []
minus = []
result = 0

for i in range(N):
    if nums[i] < 0:
        minus.append((i, nums[i]))
    elif nums[i] > 0:
        plus.append((i, nums[i]))

m = 0
p = 0

while m < len(minus):
    midx, minus_num = minus[m]

    while minus_num:
        pidx, plus_num = plus[p]
        
        if minus_num + plus_num > 0:
            result += -minus_num * abs(midx - pidx)
            plus[p] = (pidx, plus_num + minus_num)
            minus_num = 0
        else:
            result += plus_num * abs(midx - pidx)
            minus_num += plus_num
            p += 1

    m += 1

print(result)
